Kernel/VM/Range.cpp

```diff
@@ -44,5 +45,17 @@ Vector<Range, 2> Range::carve(const Range& taken) const
         parts.append({ taken.end(), end().get() - taken.end().get() });
     return parts;
 }
+Range Range::intersect(const Range& other) const
+{
+    if (*this == other) {
+        return *this;
+    }
+    auto new_base = max(base(), other.base());
+    auto new_end = max(min(end(), other.end()), base());
```

Comment:
This will crash sometimes.

Let `this` be the Range from `0x1000` to `0x2000`, and let `other` be the Range from `0x4000` to `0x7000`.
Then your code computes:
```
new_base = max(0x1000, 0x4000) = 0x4000
new_end = max(min(0x2000, 0x7000), 0x1000) = max(0x2000, 0x1000) = 0x2000
VERIFY(false);
```
Oops!

Here's a tip for life: If something is symmetric in mathematics (for example, computing the intersection of two intervals, "A intersect B" = "B intersect A"), then code that computes this symmetric operation should also be symmetric. That's not a hard rule, but it certainly raises an eyebrow if the code looks asymmetric. Case in point: `auto new_base = max(base(), other.base());` is obviously symmetric: If "this" and "other" are swapped, it obviously should produce the same result. Same for `min(end(), other.end())`. But `max(…, base());`?

This can be fixed, yes, but the only use is an incomplete micro-optimization in `find_regions_intersecting` as I already explained. (Either there is a much better one when it's sorted, or it cannot possibly save any significant amount of time.)

Reply:
Ok, I fixed that, and I realized, with a zero intersection, we would Crash, so the Verify should be  a `<=`

And I haven't made that function for preformance sake, but for readability and ease of use,
I think its easier to understand when I say that the intersection is eaqual, than saying are the bounds contained, or to calculate the intersection size at onepoint it gets down to an understandable oneliner

Reply:
So what exactly does the function assume now?

- If the Ranges are "very" disjoint, it crashes in `VERIFY`.
- If the Ranges are disjoint but "touch" (i.e. `end() == other.base()`), it passes, and (uhh, see below)

So does the function assume they intersect or not?

Also, there's a typo in the "return" line: `new_end - end()` should probably be `new_end - new_base`.

Reply:
So it should always return actually even when the ranges are very disjoint all values get clamped to end(), 
if I'm not mistaken and thenn theres a 0-Range being returned...

And thanks for hinting at the typo, it would probably cause a very weird bug somewhere else...

Reply:
And should we check for page alignment like carve does?

Reply:
I've changed it, that incase of no match it just returns an invalid region

Reply:
Yup, I think this code works. I would have written it differently, but it works, so I don't complain :D